src/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Windows.cs

```diff
@@ -126,11 +122,138 @@ public Task WaitForConnectionAsync(CancellationToken cancellationToken)
                 cancellationToken.CanBeCanceled ? new IOCancellationHelper(cancellationToken) : null);
         }
 
+        [SecurityCritical]
+        public void Disconnect()
+        {
+            CheckDisconnectOperations();
+
+            // Disconnect the pipe.
+            if (!Interop.mincore.DisconnectNamedPipe(InternalHandle))
+            {
+                throw Win32Marshal.GetExceptionForLastWin32Error();
+            }
+
+            State = PipeState.Disconnected;
+        }
+
+        // Gets the username of the connected client.  Not that we will not have access to the client's 
+        // username until it has written at least once to the pipe (and has set its impersonationLevel 
+        // argument appropriately). 
+        [SecurityCritical]
+        public String GetImpersonationUserName()
+        {
+            CheckWriteOperations();
+
+            StringBuilder userName = new StringBuilder(Interop.CREDUI_MAX_USERNAME_LENGTH + 1);
+
+            if (!Interop.mincore.GetNamedPipeHandleState(InternalHandle, Interop.NULL, Interop.NULL,
+                Interop.NULL, Interop.NULL, userName, userName.Capacity))
+            {
+                WinIOError(Marshal.GetLastWin32Error());
+            }
+
+            return userName.ToString();
+        }
+
+        // -----------------------------
+        // ---- PAL layer ends here ----
+        // -----------------------------
+
+        [SecurityCritical]
+        private unsafe static readonly IOCompletionCallback s_WaitForConnectionCallback =
+            new IOCompletionCallback(NamedPipeServerStream.AsyncWaitForConnectionCallback);
+
+#if RunAs
```

Comment:
I actually don't know... I didn't muck with this code, just moved it around in the file. (That's true for almost all of the code in the *.Windows.cs files.)